src/JsonRpcEngine.ts

```diff
@@ -296,18 +360,45 @@ export class JsonRpcEngine extends SafeEventEmitter {
         `Must specify a string method. Received: ${typeof callerReq.method}`,
         { request: callerReq },
       );
-      return callback(error, { id: callerReq.id, jsonrpc: '2.0', error });
+
+      if (this._notificationHandler && !isJsonRpcRequest(callerReq)) {
+        // Do not reply to notifications, even if they are malformed.
+        return callback(null);
+      }
+
+      return callback(error, {
+        // Typecast: This could be a notification, but we want to access the
+        // `id` even if it doesn't exist.
+        id: (callerReq as JsonRpcRequest<unknown>).id ?? null,
+        jsonrpc: '2.0',
+        error,
+      });
     }
 
-    const req: JsonRpcRequest<unknown> = { ...callerReq };
+    // Handle notifications.
+    // We can't use isJsonRpcNotification here because that narrows callerReq to
```

Comment:
I am not 100% sure on this, but my guess is that it's because there is no correlation between `this._notificationHandler` and the type of `callerReq`. If there is a way to tell TypeScript that when `this_.notificationHandler` is given then all requests _must_ be notifications then this could work, but I'm not sure quite how to do that. This seems sufficient. I appreciate the comment!